materialize-sql/type_mapping.go

```diff
@@ -142,379 +144,169 @@ type MappedType struct {
-// per the endpoint's requirements.
-func StringCastConverter(fn func(string) (interface{}, error)) ElementConverter {
-	return func(te tuple.TupleElement) (interface{}, error) {
-		switch tt := te.(type) {
-		case string:
-			return fn(tt)
-		default:
-			return te, nil
-		}
+// MapStatic creates a ProjectionMapper that always returns the same DDL and
+// optionally sets an ElementConverter.
+func MapStatic(ddl string, converter ...ElementConverter) MapProjectionFn {
+	var c ElementConverter
+	if converter != nil {
+		c = converter[0]
 	}
-}
-
-func Compose(upper ElementConverter, lower ElementConverter) ElementConverter {
-	return func(te tuple.TupleElement) (interface{}, error) {
-		var a, err = lower(te)
-		if err != nil {
-			return nil, err
-		}
 
-		return upper(a)
+	return func(p *Projection) (string, ElementConverter) {
+		return ddl, c
 	}
 }
 
-// StdByteArrayToStr builds an ElementConverter that converts []byte to string.
-func StdByteArrayToStr(te tuple.TupleElement) (interface{}, error) {
-	switch tt := te.(type) {
-	case []byte:
-		if tt == nil {
-			return nil, nil
-		}
-		return string(tt), nil
-	case json.RawMessage:
-		if tt == nil {
-			return nil, nil
+// MapPrimaryKey specifies an alternate ProjectionMapper for the the column if
+// it is a primary key. This is useful for cases where specific databases
+// require certain variations of a type for primary keys. For example, MySQL
+// does not accept TEXT as a primary key, but a VARCHAR with specified size
+// works.
+func MapPrimaryKey(pkMapper, delegate MapProjectionFn) MapProjectionFn {
+	return func(p *Projection) (string, ElementConverter) {
+		if p.IsPrimaryKey {
+			return pkMapper(p)
 		}
-		return string(tt), nil
-	default:
-		return te, nil
+		return delegate(p)
 	}
 }
 
-// StdStrToInt builds an ElementConverter that attempts to convert a string into a big.Int. Strings
-// formatted as integers are often larger than the maximum that can be represented by an int64. The
-// concrete type of the return value is a big.Int which will be represented correctly as JSON, but
-// may not be directly usable by a SQL driver as a parameter.
-func StdStrToInt() ElementConverter {
-	return StringCastConverter(func(str string) (interface{}, error) {
-		// Strings ending in a 0 decimal part like "1.0" or "3.00" are considered valid as integers
-		// per JSON specification so we must handle this possibility here. Anything after the
-		// decimal is discarded on the assumption that Flow has validated the data and verified that
-		// the decimal component is all 0's.
-		if idx := strings.Index(str, "."); idx != -1 {
-			str = str[:idx]
-		}
-
-		var i big.Int
-		out, ok := i.SetString(str, 10)
-		if !ok {
-			return nil, fmt.Errorf("could not convert %q to big.Int", str)
-		}
-		return out, nil
-	})
+type StringMappings struct {
+	Fallback        MapProjectionFn
+	WithFormat      map[string]MapProjectionFn
+	WithContentType map[string]MapProjectionFn
```

Comment:
A comment explaining what ContentType means would be useful here, it took me some research to find out what it is

Reply:
Added some additional comments here!